x/proof/keeper/msg_server_create_claim.go

```diff
@@ -85,13 +91,49 @@ func (k msgServer) CreateClaim(
 		RootHash:        msg.GetRootHash(),
 	}
 
-	// TODO_BLOCKER(@Olshansk): check if this claim already exists and return an
-	// appropriate error in any case where the supplier should no longer be able
-	// to update the given proof.
+	_, isExistingClaim = k.Keeper.GetClaim(ctx, claim.GetSessionHeader().GetSessionId(), claim.GetSupplierAddress())
+
 	k.Keeper.UpsertClaim(ctx, claim)
 
 	logger.Info("created new claim")
 
+	// NB: Don't return these error, it should not prevent the MsgCreateClaimResopnse
+	// from being returned. Instead, it will be attached as an "error" label to any
+	// metrics tracked in this function.
+	// TODO_IMPROVE: While this will surface the error in metrics, it will also cause
+	// any counters not to be incremented even though a new claim might have been inserted.
+	numRelays, err = claim.GetNumRelays()
+	numComputeUnits, err = claim.GetNumComputeUnits()
+
+	// Emit the appropriate event based on whether the claim was created or updated.
+	var claimUpsertEvent proto.Message
+	switch isExistingClaim {
+	case true:
+		claimUpsertEvent = proto.Message(
+			&types.EventClaimUpdated{
+				Claim:           &claim,
+				NumRelays:       numRelays,
+				NumComputeUnits: numComputeUnits,
+			},
+		)
+	case false:
+		claimUpsertEvent = proto.Message(
+			&types.EventClaimCreated{
+				Claim:           &claim,
+				NumRelays:       numRelays,
+				NumComputeUnits: numComputeUnits,
+			},
+		)
+	}
+
+	sdkCtx := cosmostypes.UnwrapSDKContext(ctx)
+	// NB: Don't return this error, it should not prevent the MsgCreateClaimResopnse
+	// from being returned. Instead, it will be attached as an "error" label to any
+	// metrics tracked in this function.
+	// TODO_IMPROVE: While this will surface the error in metrics, it will also cause
+	// any counters not to be incremented even though a new claim might have been inserted.
+	err = sdkCtx.EventManager().EmitTypedEvent(claimUpsertEvent)
```

Comment:
```suggestion
	emitEventErr := sdkCtx.EventManager().EmitTypedEvent(claimUpsertEvent)
	
	// NB: Don't return these error, it should not prevent the MsgCreateClaimResopnse
	// from being returned. Instead, it will be attached as an "error" label to any
	// metrics tracked in this function.
	err = errors.Join(numRelaysErr, numComputeUnitsErr, emitEventErr)
```